analysis/common_variables.py

```diff
@@ -58,30 +58,66 @@
             },
         },
     ),
-    imd=patients.categorised_as(
+    imd=patients.address_as_of(
+        "index_date",
+        returning="index_of_multiple_deprivation",
+        round_to_nearest=100,
+        return_expectations={
+            "rate": "universal",
+            "category": {
+                "ratios": {
+                    "-1": 0.05,
+                    "100": 0.19,
+                    "200": 0.19,
+                    "300": 0.19,
+                    "400": 0.19,
+                    "500": 0.19,
+                }
+            },
+        },
+    ),
+    imdQ5_incorrect=patients.categorised_as(
         {
-            "0": "DEFAULT",
-            "1": """index_of_multiple_deprivation >=1 AND index_of_multiple_deprivation < 32844*1/5""",
-            "2": """index_of_multiple_deprivation >= 32844*1/5 AND index_of_multiple_deprivation < 32844*2/5""",
-            "3": """index_of_multiple_deprivation >= 32844*2/5 AND index_of_multiple_deprivation < 32844*3/5""",
-            "4": """index_of_multiple_deprivation >= 32844*3/5 AND index_of_multiple_deprivation < 32844*4/5""",
-            "5": """index_of_multiple_deprivation >= 32844*4/5 AND index_of_multiple_deprivation < 32844""",
+            "Unknown": "DEFAULT",
+            "1 (most deprived)": """imd >=1 AND imd < 32844*1/5""",
+            "2": """imd >= 32844*1/5 AND imd < 32844*2/5""",
+            "3": """imd >= 32844*2/5 AND imd < 32844*3/5""",
+            "4": """imd >= 32844*3/5 AND imd < 32844*4/5""",
+            "5 (least deprived)": """imd >= 32844*4/5 AND imd <= 32844""",
+        },
+        return_expectations={
+            "rate": "universal",
+            "category": {
+                "ratios": {
+                    "Unknown": 0.05,
+                    "1 (most deprived)": 0.19,
+                    "2": 0.19,
+                    "3": 0.19,
+                    "4": 0.19,
+                    "5 (least deprived)": 0.19,
+                }
+            },
+        },
+    ),
+    imdQ5_correct=patients.categorised_as(
+        {
+            "Unknown": "DEFAULT",
+            "1 (most deprived)": """imd >=0 AND imd < 32844*1/5""",
+            "2": """imd >= 32844*1/5 AND imd < 32844*2/5""",
+            "3": """imd >= 32844*2/5 AND imd < 32844*3/5""",
+            "4": """imd >= 32844*3/5 AND imd < 32844*4/5""",
+            "5 (least deprived)": """imd >= 32844*4/5 AND imd <= 32844""",
```

Comment:
The maximum value of `imd` is 32800 not 32844, making the buckets 6560 not 6568.8. Using the larger maximum value misclassifies eight LSOAs per category: they will be one category lower than they should be.

Reply:
It's not quite as simple as that.

Retaining the original 32844 introduces less bias because the original group thresholds are kept and rounding-to-nearest-100 is itself an unbiased operation _over all integers_ (or reals). That is, the group-misclassification-due-to-rounding averages over infinitely many integers. A randomly selected rounded integer has the same chance of being misclassified into the group one below as into the group one above.

By contrast, changing the maximum to 32800 changes the cut-points downwards so that, over all integers, rounding makes it more likely that rounded integers are misclassified into the group one above than the group one below.

So knowing no more, sticking with 32844/5 makes more sense. But we're only dealing with finitely many IMD ranks. And a fixed number at that. So maybe the other approach is better for this specific number. So we might as well check:

```R
maximd <- 32844
maximd100 <- plyr::round_any(maximd, 100)

imd_data <-
  tibble::tibble(
    imd = 1:maximd,
    imd100 = plyr::round_any(imd, 100), # round to nearest even 100, though I think TPP do half-up rounding
    imdGmax = cut(imd, c(0,maximd*(1:5)/5), right=FALSE, include.lowest=TRUE, labels=FALSE),
    imd100Gmax = cut(imd100, c(0,maximd*(1:5)/5), right=FALSE, include.lowest=TRUE, labels=FALSE),
    imd100Gmax100 = cut(imd100, c(0,maximd100*(1:5)/5), right=FALSE, include.lowest=TRUE, labels=FALSE),
  )


with(imd_data, table("true group" = imdGmax, "with 32844/5" = imd100Gmax))
with(imd_data, table("true group" = imdGmax, "with 32800/5" = imd100Gmax100))


classification_summary <- 
  dplyr::summarise(
    imd_data,
    imd100Gmax_misclassified = sum(imd100Gmax != imdGmax),
    imd100Gmax_misclassifiedup = sum(imd100Gmax > imdGmax),
    imd100Gmax_misclassifieddown = sum(imd100Gmax < imdGmax),
    imd100Gmax100_misclassified = sum(imd100Gmax100 != imdGmax),
    imd100Gmax100_misclassifiedup = sum(imd100Gmax100 > imdGmax),
    imd100Gmax100_misclassifieddown = sum(imd100Gmax100 < imdGmax),
  )

t(classification_summary)

```

gives:

```R
> with(imd_data, table("true group" = imdGmax, "with 32844/5" = imd100Gmax))
          with 32844/5
true group    1    2    3    4    5
         1 6549   19    0    0    0
         2    0 6569    0    0    0
         3    0   12 6557    0    0
         4    0    0   43 6501   25
         5    0    0    0    0 6569
> with(imd_data, table("true group" = imdGmax, "with 32800/5" = imd100Gmax100))
          with 32800/5
true group    1    2    3    4    5
         1 6549   19    0    0    0
         2    0 6569    0    0    0
         3    0   12 6501   56    0
         4    0    0    0 6544   25
         5    0    0    0    0 6569

> t(classification_summary)
                                [,1]
imd100Gmax_misclassified          99
imd100Gmax_misclassifiedup        44
imd100Gmax_misclassifieddown      55
imd100Gmax100_misclassified      112
imd100Gmax100_misclassifiedup    100
imd100Gmax100_misclassifieddown   12
```

The first method (32844) misclassifies 99 in total, with 44 too high and 55 too low.

The second method (32800) misclassifies 112 in total, with 100 too high and 12 too low.

So 32844 wins.